### PR TITLE
Add custom Eth module that exposes additional OP related fields

### PR DESCRIPTION
**Pull request by flcl42**
## Changes

- Add OP specific fields to the receipts. They are different for deposit tx and the rest

#### What types of changes does your code introduce?

- [ ] Bugfix (a non-breaking change that fixes an issue)
- [x] New feature (a non-breaking change that adds functionality)
- [ ] Breaking change (a change that causes existing functionality not to work as expected)
- [ ] Optimization
- [ ] Refactoring
- [ ] Documentation update
- [ ] Build-related changes
- [ ] Other: _Description_

#### Requires testing

- [x] Yes
- [ ] No

#### If yes, did you write tests?

- [x] Yes
- [ ] No

#### Notes on testing

_Optional. Remove if not applicable._

#### Requires documentation update

- [x] Yes
- [x] No

### CODE REVIEW

**Review by LukaszRozmej — COMMENTED** [review]
looks a bit overengeneered, couldn't we have `OptimismEthRpcModule : IEthRpcModule` and then return results with derived classes that can have different json serialization rules?

File: src/Nethermind/Nethermind.Serialization.Rlp/ReceiptArrayStorageDecoder.cs [hunk 1]
```diff
@@ -19,12 +19,7 @@ public class ReceiptArrayStorageDecoder : IRlpStreamDecoder<TxReceipt[]>
     private static readonly IRlpValueDecoder<TxReceipt> CompactValueDecoder = Rlp.GetValueDecoder<TxReceipt>(RlpDecoderKey.Storage);
 
     public const int CompactEncoding = 127;
-    private readonly bool _useCompactEncoding = true;
-
-    public ReceiptArrayStorageDecoder(bool compactEncoding = true)
-    {
-        _useCompactEncoding = compactEncoding;
-    }
+    private readonly bool _useCompactEncoding = compactEncoding;
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Serialization.Rlp/ReceiptArrayStorageDecoder.cs:22]: this field is probably not needed now.

File: src/Nethermind/Nethermind.Optimism/Rpc/IOptimismEthRpcModule.cs [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+using Nethermind.JsonRpc;
+using Nethermind.JsonRpc.Data;
+using Nethermind.JsonRpc.Modules;
+using Nethermind.JsonRpc.Modules.Eth;
+using Nethermind.State.Proofs;
+
+namespace Nethermind.Optimism;
+
+[RpcModule(ModuleType.Eth)]
+public interface IOptimismEthRpcModule : IRpcModule
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Optimism/Rpc/IOptimismEthRpcModule.cs:21]: can this extend IEthRpcModule?

Comment (flcl42) [reply src/Nethermind/Nethermind.Optimism/Rpc/IOptimismEthRpcModule.cs:21]: Let me try to solve it via generics

File: src/Nethermind/Nethermind.Serialization.Rlp/ReceiptArrayStorageDecoder.cs (hunk 1 above)
Comment (flcl42) [reply src/Nethermind/Nethermind.Serialization.Rlp/ReceiptArrayStorageDecoder.cs:22]: If you mean we can drop non compact encoding, there a small concern: IReceiptConfig mentions that usage of compact encoding may effect RPC performance. Does it mean non compact receipts may still have users? If no - let's remove old encoding for sure

Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Serialization.Rlp/ReceiptArrayStorageDecoder.cs:22]: I mean you can use default constructor parameter `compactEncoding` and don't need the field.

File: src/Nethermind/Nethermind.Optimism/Rpc/IOptimismEthRpcModule.cs (hunk 2 above)
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Optimism/Rpc/IOptimismEthRpcModule.cs:21]: why would they be needed? You can have a return type of base class and then return derived type - it should serialize correctly in json rpc.

**Comment by flcl42** [comment]
@LukaszRozmej the latest commit a bit more complicated. But in addition to deduplication, I wanted to bring this general PoC of abstracted key data structures. Maybe we can do the same with transactions, blocks, etc later

File: src/Nethermind/Nethermind.Optimism/Rpc/IOptimismEthRpcModule.cs (hunk 2 above)
Comment (flcl42) [reply src/Nethermind/Nethermind.Optimism/Rpc/IOptimismEthRpcModule.cs]: We can use IResultWrapper, but we will lose type information. Which may be useful for tests, documentation generation and less convenient for devs.

File: src/Nethermind/Nethermind.Serialization.Rlp/ReceiptArrayStorageDecoder.cs (hunk 1 above)
Comment (flcl42) [reply src/Nethermind/Nethermind.Serialization.Rlp/ReceiptArrayStorageDecoder.cs:22]: Oh indeed

File: src/Nethermind/Nethermind.Optimism/Rpc/OptimismEthRpcModule.cs [hunk 3]
```diff
@@ -2,151 +2,109 @@
...
 using Nethermind.Serialization.Rlp;
-using Nethermind.State.Proofs;
+using Nethermind.State;
 using Nethermind.TxPool;
+using Nethermind.Wallet;
 
 namespace Nethermind.Optimism;
 
-public class OptimismEthRpcModule : IEthRpcModule
+public class OptimismEthRpcModule : EthRpcModule, IOptimismEthRpcModule
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Optimism/Rpc/OptimismEthRpcModule.cs:31]: Do we benefit from inheritance over composition?

**Review by deffrian — APPROVED** [review]

**Review by LukaszRozmej — APPROVED** [review]
I kind of would prefer composition for the module, but it is fine.

**Comment by flcl42** [comment]
> I kind of would prefer composition for the module, but it is fine.

@LukaszRozmej I don't mind, but why?

**Comment by LukaszRozmej** [comment]
> > I kind of would prefer composition for the module, but it is fine.
> 
> @LukaszRozmej I don't mind, but why?

Composition is always more flexible

**Pull request merged**